src/crimson/os/seastore/nvmedevice/nvmedevice.h

```diff
@@ -0,0 +1,329 @@
+    struct
+    {
+      uint8_t raw[4096];
+    };
+    struct
+    {
+      uint8_t unused[64];
+      uint16_t npwg;
+      uint16_t npwa;
+    };
+  };
+};
+
+using NVMePassThroughCommand = nvme_passthru_cmd;
+
+using read_ertr = crimson::errorator<
+  crimson::ct_error::input_output_error,
+  crimson::ct_error::invarg,
+  crimson::ct_error::enoent,
+  crimson::ct_error::erange>;
+
+using write_ertr = crimson::errorator<
+  crimson::ct_error::input_output_error,
+  crimson::ct_error::invarg,
+  crimson::ct_error::ebadf,
+  crimson::ct_error::enospc>;
+
+using open_ertr = crimson::errorator<
+  crimson::ct_error::input_output_error,
+  crimson::ct_error::invarg,
+  crimson::ct_error::enoent>;
+
+using nvme_command_ertr = crimson::errorator<
+  crimson::ct_error::input_output_error>;
+
+struct io_context_t {
+  iocb cb;
+  bool done = false;
+};
+
+/*
+ * Interface between NVMe SSD and its user.
+ *
+ * NVMeBlockDevice provides not only the basic APIs for IO, but also helper APIs
+ * to accelerate SSD IO performance and reduce system overhead. By aggresively
+ * utilizing and abstract useful features of latest NVMe SSD, it helps user ease
+ * to get high performance of NVMe SSD and low system overhead.
+ *
+ * Various implementations with different interfaces such as POSIX APIs, Seastar,
+ * and SPDK, are available.
+ */
+class NVMeBlockDevice {
+protected:
+  uint64_t size = 0;
+  uint64_t block_size = 0;
+
+  uint64_t write_granularity = 4096;
+  uint64_t write_alignment = 4096;
+
+public:
+  NVMeBlockDevice() {}
+  virtual ~NVMeBlockDevice() = default;
+
+  template <typename T>
+  static NVMeBlockDevice *create() {
+    return new T();
+  }
+
+  /*
+   * Service NVMe device relative size
+   *
+   * size : total size of device in byte.
+   *
+   * block_size : IO unit size in byte. Caller should follow every IO command
+   * aligned with block size.
+   *
+   * preffered_write_granularity(PWG), preffered_write_alignment(PWA) : IO unit
+   * size for write in byte. Caller should request every write IO sized multiple
+   * times of PWG and aligned starting address by PWA. Available only if NVMe
+   * Device supports NVMe protocol 1.4 or later versions.
+   */
+  uint64_t get_size() const { return size; }
+  uint64_t get_block_size() const { return block_size; }
+  uint64_t get_preffered_write_granularity() const { return write_granularity; }
+  uint64_t get_preffered_write_alignment() const { return write_alignment; }
+
+  virtual read_ertr::future<> read(
+    uint64_t offset,
+    bufferptr &bptr) = 0;
+
+  /*
+   * Multi-stream write
+   *
+   * Give hint to device about classification of data whose life time is similar
+   * with each other. Data with same stream value will be managed together in
+   * SSD for better write performance.
+   */
+  virtual write_ertr::future<> write(
+    uint64_t offset,
+    bufferptr &bptr,
+    uint16_t stream = 0) = 0;
+
+  // TODO
+  virtual int discard(uint64_t offset, uint64_t len) { return 0; }
```

Comment:
This would need to be a future/errorator, right?

Reply:
Right